File: backtest-service/src/main/resources/scripts/python_strategy_runner.py
#!/usr/bin/env python3
"""
Python Strategy Runner Bridge
==============================
Reads a JSON request from stdin, executes the user-provided Python strategy
code, calls the specified entrypoint, and prints a JSON result to stdout.

Input JSON structure (stdin):
{
  "code": "def on_bar(ctx): ...",
  "entrypoint": "on_bar",
  "ctx": {
    "params": { ... },
    "indicators": { ... },
    "position": { "qty": 0 },
    "bar": { "symbol": "...", "close": 100.0, "timestamp": "..." }
  }
}

Output JSON structure (stdout) on success:
{
  "success": true,
  "result": { "action": "HOLD" | "BUY" | "SELL", "qty": 100 }
}

Output JSON structure (stdout) on error:
{
  "success": false,
  "error": {
    "type": "SYNTAX_ERROR" | "ENTRYPOINT_NOT_FOUND" | "EXECUTION_ERROR" | "INVALID_RETURN",
    "message": "..."
  }
}
"""

import json
import sys
import traceback


def main():
    request_raw = sys.stdin.read()
    if not request_raw:
        _fail("EXECUTION_ERROR", "No input received on stdin")
        return

    try:
        request = json.loads(request_raw)
    except json.JSONDecodeError as e:
        _fail("EXECUTION_ERROR", f"Invalid JSON input: {e}")
        return

    code = request.get("code", "")
    if not code:
        _fail("EXECUTION_ERROR", "Missing 'code' field")
        return

    entrypoint = request.get("entrypoint", "on_bar")
    ctx = request.get("ctx", {})

    # 1. Compile user code (catches syntax errors)
    try:
        compiled = compile(code, "<strategy_code>", "exec")
    except SyntaxError as e:
        _fail("SYNTAX_ERROR", f"Python syntax error: {e}")
        return
    except Exception as e:
        _fail("SYNTAX_ERROR", f"Compilation error: {e}")
        return

    # 2. Execute in isolated namespace
    namespace = {}
    try:
        exec(compiled, namespace)
    except Exception as e:
        _fail("EXECUTION_ERROR", f"Error during code execution: {e}\n{traceback.format_exc()}")
        return

    # 3. Check entrypoint exists
    if entrypoint not in namespace:
        found = [k for k in namespace.keys() if not k.startswith("__")]
        _fail(
            "ENTRYPOINT_NOT_FOUND",
            f"Entrypoint '{entrypoint}' not found in user code. "
            f"Available definitions: {found}"
        )
        return

    entry_fn = namespace[entrypoint]
    if not callable(entry_fn):
        _fail(
            "ENTRYPOINT_NOT_FOUND",
            f"'{entrypoint}' is not callable (type: {type(entry_fn).__name__})"
        )
        return

    # 4. Call entrypoint(ctx)
    try:
        result = entry_fn(ctx)
    except Exception as e:
        _fail(
            "EXECUTION_ERROR",
            f"Error executing {entrypoint}(ctx): {e}\n{traceback.format_exc()}"
        )
        return

    # 5. Validate result
    if not isinstance(result, dict):
        _fail(
            "INVALID_RETURN",
            f"Entrypoint returned {type(result).__name__}, expected dict. "
            f"Return value: {result}"
        )
        return

    action = result.get("action")
    if action not in ("HOLD", "BUY", "SELL"):
        _fail(
            "INVALID_RETURN",
            f"Invalid action '{action}'. Must be one of: HOLD, BUY, SELL. "
            f"Return value: {result}"
        )
        return

    qty = result.get("qty")
    if action in ("BUY", "SELL") and qty is not None:
        if not isinstance(qty, (int, float)):
            _fail(
                "INVALID_RETURN",
                f"qty must be a number for {action}, got {type(qty).__name__}: {qty}"
            )
            return
        if qty <= 0:
            _fail(
                "INVALID_RETURN",
                f"qty must be positive for {action}, got: {qty}"
            )
            return

    # 6. Success
    _succeed({"action": action, "qty": qty})


def _succeed(result_dict):
    output = {"success": True, "result": result_dict}
    json.dump(output, sys.stdout, ensure_ascii=False)
    sys.stdout.flush()


def _fail(error_type, message):
    output = {"success": False, "error": {"type": error_type, "message": message}}
    json.dump(output, sys.stdout, ensure_ascii=False)
    sys.stdout.flush()


if __name__ == "__main__":
    main()

File: backtest-service/src/main/java/ai/jzhu/trading/backtest/domain/service/PythonStrategyRunner.java
package ai.jzhu.trading.backtest.domain.service;

import ai.jzhu.trading.backtest.domain.model.PythonStrategyExecutionRequest;
import ai.jzhu.trading.backtest.domain.model.PythonStrategyExecutionResult;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Executes Python strategy code via a subprocess pipe.
 *
 * <p>The runner writes a JSON request to stdin of the Python bridge script,
 * then reads a JSON response from stdout. Timeout and error handling are
 * applied at this boundary.
 *
 * <p>This is NOT a sandbox — it relies on OS-level process isolation only.
 * Full sandboxing (e.g., restricted user, seccomp, gVisor) is out of scope
 * for this step.
 */
@Component
public class PythonStrategyRunner {

    private static final Logger log = LoggerFactory.getLogger(PythonStrategyRunner.class);

    /** Maximum wall-clock time (milliseconds) for a single on_bar call. */
    private static final long DEFAULT_TIMEOUT_MS = 5_000;

    private final ObjectMapper objectMapper;
    private final String pythonCommand;
    private final String bridgeScriptPath;
    private final long timeoutMs;

    public PythonStrategyRunner() {
        this.objectMapper = new ObjectMapper();
        this.pythonCommand = resolvePythonCommand();
        this.bridgeScriptPath = resolveBridgeScriptPath();
        this.timeoutMs = DEFAULT_TIMEOUT_MS;
    }

    // Constructor exposed for testing (timeout injection)
    PythonStrategyRunner(String pythonCommand, String bridgeScriptPath, long timeoutMs) {
        this.objectMapper = new ObjectMapper();
        this.pythonCommand = pythonCommand;
        this.bridgeScriptPath = bridgeScriptPath;
        this.timeoutMs = timeoutMs;
    }

    /**
     * Execute one on_bar(ctx) invocation synchronously.
     *
     * @param code       Python source code.
     * @param entrypoint Function name to call.
     * @param ctx        Context map passed to the function.
     * @return Execution result (success or structured error).
     */
    public PythonStrategyExecutionResult execute(
            String code,
            String entrypoint,
            Map<String, Object> ctx
    ) {
        var request = new PythonStrategyExecutionRequest(code, entrypoint, ctx);
        return execute(request);
    }

    /**
     * Execute one on_bar(ctx) invocation synchronously.
     */
    public PythonStrategyExecutionResult execute(PythonStrategyExecutionRequest request) {
        try {
            // Build the input JSON
            Map<String, Object> input = Map.of(
                    "code", request.code(),
                    "entrypoint", request.entrypoint(),
                    "ctx", request.ctx()
            );
            String inputJson = objectMapper.writeValueAsString(input);

            // Start the Python subprocess
            ProcessBuilder pb = new ProcessBuilder(pythonCommand, bridgeScriptPath);
            pb.redirectErrorStream(false);

            Process process = pb.start();

            // Write input to stdin
            try (OutputStreamWriter writer = new OutputStreamWriter(process.getOutputStream(), StandardCharsets.UTF_8)) {
                writer.write(inputJson);
                writer.flush();
            }

            // Read stdout and stderr concurrently BEFORE waitFor to prevent pipe buffer deadlock.
            // If the Python process writes enough output to fill the OS pipe buffer (~64KB on Linux),
            // it will block on write(). If we waitFor() before draining the pipes, both sides hang.
            StringBuilder stdout = new StringBuilder();
            StringBuilder stderr = new StringBuilder();

            Thread stdoutReader = new Thread(() -> {
                try (BufferedReader reader = new BufferedReader(
                        new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        stdout.append(line);
                    }
                } catch (java.io.IOException ignored) {
                    // stream closed after process exits — expected
                }
            });
            stdoutReader.setDaemon(true);

            Thread stderrReader = new Thread(() -> {
                try (BufferedReader reader = new BufferedReader(
                        new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        stderr.append(line);
                    }
                } catch (java.io.IOException ignored) {
                    // stream closed after process exits — expected
                }
            });
            stderrReader.setDaemon(true);

            stdoutReader.start();
            stderrReader.start();

            // Wait with timeout (pipes are being drained concurrently)
            boolean finished = process.waitFor(timeoutMs, TimeUnit.MILLISECONDS);

            // Ensure readers finish consuming any remaining buffered output
            stdoutReader.join(500);
            stderrReader.join(500);

            if (!finished) {
                process.destroyForcibly();
                log.warn("Python strategy execution timed out after {}ms", timeoutMs);
                return PythonStrategyExecutionResult.failure(
                        "TIMEOUT",
                        "Python strategy execution timed out after " + timeoutMs + "ms"
                );
            }

            int exitCode = process.exitValue();
            if (exitCode != 0 && stdout.length() == 0) {
                String stderrMsg = stderr.length() > 0 ? stderr.toString() : "Unknown error (exit code " + exitCode + ")";
                log.warn("Python strategy runner exited with code {}: {}", exitCode, stderrMsg);
                return PythonStrategyExecutionResult.failure("EXECUTION_ERROR", stderrMsg);
            }

            String outputStr = stdout.toString().trim();
            if (outputStr.isEmpty()) {
                return PythonStrategyExecutionResult.failure(
                        "EXECUTION_ERROR",
                        "Python runner produced no output (exit code " + exitCode + ")"
                );
            }

            // Parse JSON response from bridge script
            Map<String, Object> response = objectMapper.readValue(
                    outputStr,
                    new TypeReference<Map<String, Object>>() {
                    }
            );

            boolean bridgeSuccess = Boolean.TRUE.equals(response.get("success"));
            if (!bridgeSuccess) {
                @SuppressWarnings("unchecked")
                Map<String, String> error = (Map<String, String>) response.get("error");
                String errorType = error != null ? error.get("type") : "UNKNOWN";
                String errorMsg = error != null ? error.get("message") : "Unknown error";
                return PythonStrategyExecutionResult.failure(errorType, errorMsg);
            }

            @SuppressWarnings("unchecked")
            Map<String, Object> result = (Map<String, Object>) response.get("result");
            String action = result != null ? (String) result.get("action") : "HOLD";
            Double qty = result != null && result.get("qty") != null
                    ? ((Number) result.get("qty")).doubleValue()
                    : null;

            // Final validation in Java side
            if (action == null || (!action.equals("HOLD") && !action.equals("BUY") && !action.equals("SELL"))) {
                return PythonStrategyExecutionResult.failure(
                        "INVALID_RETURN",
                        "Invalid action: " + action
                );
            }

            return PythonStrategyExecutionResult.success(action, qty);

        } catch (Exception e) {
            log.error("Unexpected error in Python strategy runner", e);
            return PythonStrategyExecutionResult.failure(
                    "EXECUTION_ERROR",
                    "Unexpected runner error: " + e.getMessage()
            );
        }
    }

    /**
     * Resolve the Python command. In Docker container 'python3' is the canonical name.
     */
    private static String resolvePythonCommand() {
        // In CI / Docker environment "python3" is the canonical name.
        // Fallback to "python" for environments where python3 does not exist.
        return "python3";
    }

    /**
     * Locate the bridge script relative to the classpath.
     * When running from a JAR, it will be on the classpath as a resource.
     * When running from IDE / Maven, it will be under target/classes/scripts/.
     */
    static String resolveBridgeScriptPath() {
        // The script is bundled in the JAR under scripts/python_strategy_runner.py
        // We extract it to a temp file to run it via subprocess.
        try {
            java.io.InputStream in = PythonStrategyRunner.class.getClassLoader()
                    .getResourceAsStream("scripts/python_strategy_runner.py");
            if (in == null) {
                throw new IllegalStateException(
                        "scripts/python_strategy_runner.py not found on classpath"
                );
            }
            java.nio.file.Path tempFile = java.nio.file.Files.createTempFile("python_strategy_runner_", ".py");
            java.nio.file.Files.copy(in, tempFile, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
            tempFile.toFile().deleteOnExit();
            return tempFile.toAbsolutePath().toString();
        } catch (Exception e) {
            throw new IllegalStateException("Failed to locate python_strategy_runner.py", e);
        }
    }
}